phac-nml/irida-next
# Namespace PUID

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in persistent unique identifiers `puid` to the `Namespace` model and removes the `puid` from the `Project` model

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

**Group PUID**

![image](https://github.com/phac-nml/irida-next/assets/25466211/edbc0648-84cd-478b-a9ac-fbb48ef9f090)


## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Run the migrations and in the `rails console` verify that all the `Namespace` records have a `puid` attribute with `Namespace.where(puid: nil)`
2. Verify that the `Project` records no longer have a `puid` attribute
3. Sign in as `admin@email.com`
4. Verify that the `puid` displays on the show page for a project
5. Verify that the `puid` displays on the show page for a group
6. Create a new Group
7. In rails console verify this group has a `puid` with `Group.last` and `Namespace.last`
8. Create a new Project
9. In rails console verify this project doesn't have a `puid` but it's namespace does with `Project.last` and `Namespace.last`
10. In the browser go to `localhost:3000/graphiql`
11. Query a known group by the group `puid`
```
group(puid: "PUID_HERE") {
  id,
  puid,
  name
}
```
12. Query a known group via a namespace by it's `puid`
```
namespace(puid: "PUID_HERE") {
  id,
  puid,
  name
}
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
